### PR TITLE
Add role assignments for ingress application gateway and corresponding example

### PR DESCRIPTION
**Pull request by lonegunmanb**
## Describe your changes

This pr added `azurerm_role_assignment` resources so the existing application gateway could work immediately. It also added a working example.

This pr contains breaking change.

## Issue number

#223 

## Checklist before requesting a review
- [ ] The pr title can be used to describe what this pr did in `CHANGELOG.md` file
- [ ] I have executed pre-commit on my machine
- [ ] I have passed pr-check on my machine

Thanks for your cooperation!

### CODE REVIEW

**Comment by lonegunmanb** [comment]
Hi @zioproto, would you please give this pr a review? Thanks!

File: log_analytics.tf [hunk 1]
```diff
@@ -0,0 +1,65 @@
+resource "azurerm_log_analytics_workspace" "main" {
```
Comment (zioproto) [inline log_analytics.tf:1]: How this change about `azurerm_log_analytics_workspace` is related with the role assignments for ingress ? Should this go to a different PR ?

File: main.tf [hunk 2]
```diff
@@ -261,9 +261,7 @@ resource "azurerm_kubernetes_cluster" "main" {
     }
   }
   dynamic "api_server_access_profile" {
-    for_each = var.api_server_authorized_ip_ranges != null || var.api_server_subnet_id != null ? [
-      "api_server_access_profile"
-    ] : []
+    for_each = var.api_server_authorized_ip_ranges != null || var.api_server_subnet_id != null ? ["api_server_access_profile"] : []
```
Comment (zioproto) [inline main.tf:264]: Is this a change in formatting that was applied automatically by `terraform fmt` ?

File: main.tf [hunk 3]
```diff
@@ -437,9 +435,7 @@ resource "azurerm_kubernetes_cluster" "main" {
     service_cidr        = var.net_profile_service_cidr
 
     dynamic "load_balancer_profile" {
-      for_each = var.load_balancer_profile_enabled && var.load_balancer_sku == "standard" ? [
-        "load_balancer_profile"
-      ] : []
+      for_each = var.load_balancer_profile_enabled && var.load_balancer_sku == "standard" ? ["load_balancer_profile"] : []
```
Comment (zioproto) [inline main.tf:438]: Change in formatting, why ?

**Comment by zioproto** [comment]
This PR fixes the roles assignments when the Application Gateway is created by user with an external to the module Terraform resource.

The scenario depicted in issue #223 is when we expect the AGIC controller running in the cluster to create the Application Gateway on behalf of the user.

I tried with the following config:

```
 ingress_application_gateway_enabled   = true
 ingress_application_gateway_subnet_id = module.network-westeurope.vnet_subnets[2]
 ingress_application_gateway_name = "aks-agw-westeurope"
```

And I still see in my `kube-system` namespace the Pod `ingress-appgw-deployment-c5d78b4f8-6dgtx` failing to create the Application Gateway

```
ingress-appgw-deployment-c5d78b4f8-6dgtx ingress-appgw-container I1123 09:54:13.013152       1 client.go:314] Deploying Gateway
ingress-appgw-deployment-c5d78b4f8-6dgtx ingress-appgw-container I1123 09:54:13.030598       1 client.go:321] Using resource group: MC_istio-aks-westeurope_westeurope-aks_westeurope
ingress-appgw-deployment-c5d78b4f8-6dgtx ingress-appgw-container I1123 09:54:13.030631       1 client.go:324] Starting ARM template deployment: aks-agw-westeurope
ingress-appgw-deployment-c5d78b4f8-6dgtx ingress-appgw-container F1123 09:54:44.072080       1 main.go:168] Failed in deploying Application GatewayCode="DeploymentFailed" Message="At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details." Details=[{"code":"Conflict","message":"{\r\n  \"error\": {\r\n    \"code\": \"ApplicationGatewayInsufficientPermissionOnSubnet\",\r\n    \"message\": \"Client with object id d0ba9abb-427c-450b-9bfb-30689a03c3d4 does not have permission on the Virtual Network resource /subscriptions/9b70acd9-975f-44ba-bad6-255a2c8bda37/resourceGroups/istio-aks-westeurope/providers/Microsoft.Network/virtualNetworks/istio-aks-westeurope/subnets/user to perform action Microsoft.Network/virtualNetworks/subnets/join/action. For details on the required permissions, please visit https://aka.ms/agsubnetjoin.\",\r\n    \"details\": []\r\n  }\r\n}"}]
```

**Review by zioproto — CHANGES_REQUESTED** [review]

File: locals.tf [hunk 4]
```diff
@@ -10,8 +20,9 @@ locals {
     (contains(["rapid", "stable", "node-image"], var.automatic_channel_upgrade) && var.kubernetes_version == null && var.orchestrator_version == null)
   )
   # Abstract the decision whether to create an Analytics Workspace or not.
-  create_analytics_solution  = var.log_analytics_workspace_enabled && var.log_analytics_solution == null
-  create_analytics_workspace = var.log_analytics_workspace_enabled && var.log_analytics_workspace == null
+  create_analytics_solution                       = var.log_analytics_workspace_enabled && var.log_analytics_solution == null
+  create_analytics_workspace                      = var.log_analytics_workspace_enabled && var.log_analytics_workspace == null
+  create_role_assignments_for_application_gateway = try(var.application_gateway_for_ingress.create_role_assignments, false)
```
Comment (zioproto) [inline locals.tf:25]: The role assignments should be created if `var.ingress_application_gateway_enabled = true`

**Comment by zioproto** [comment]
In my testing scenario each nodepool has a dedicated subnet. I have a `system` and a `user` nodepool in a `system` and a `user` subnet:

https://github.com/zioproto/istio-aks-example/blob/main/multicluster-istio-on-aks/nodepools.tf
https://github.com/zioproto/istio-aks-example/blob/main/multicluster-istio-on-aks/network.tf#L8
https://github.com/zioproto/istio-aks-example/blob/main/multicluster-istio-on-aks/main.tf#L15

When the AGIC creates the Application Gateway I see the error `ApplicationGatewayInsufficientPermissionOnSubnet`.

```
ingress-appgw-deployment-c5d78b4f8-6dgtx ingress-appgw-container F1123 10:09:20.881644       1 main.go:168] Failed in deploying Application GatewayCode="DeploymentFailed" Message="At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details." Details=[{"code":"Conflict","message":"{\r\n  \"error\": {\r\n    \"code\": \"ApplicationGatewayInsufficientPermissionOnSubnet\",\r\n    \"message\": \"Client with object id d0ba9abb-427c-450b-9bfb-30689a03c3d4 does not have permission on the Virtual Network resource /subscriptions/9b70acd9-975f-44ba-bad6-255a2c8bda37/resourceGroups/istio-aks-westeurope/providers/Microsoft.Network/virtualNetworks/istio-aks-westeurope/subnets/user to perform action Microsoft.Network/virtualNetworks/subnets/join/action. For details on the required permissions, please visit https://aka.ms/agsubnetjoin.\",\r\n    \"details\": []\r\n  }\r\n}"}]
```

The problem seems to be there only for the `user` subnet.

**Comment by zioproto** [comment]
Next steps:

* The `application_gateway_for_ingress` object should have Optional `ID`.
* The old variables `ingress_application_gateway_enabled` `ingress_application_gateway_name` `ingress_application_gateway_subnet_cidr` `ingress_application_gateway_subnet_id` will be removed and will go into the `application_gateway_for_ingress` object
* The role assignments will be created both in the scenario where the customer brings his own application gateway with an existing ID, and in the scenario where the customer only gives a name, and the application gateway is created by the AGIC Pod in the AKS control plane.

File: main.tf [hunk 5]
```diff
@@ -571,6 +571,10 @@ resource "azurerm_kubernetes_cluster" "main" {
       (var.client_id == "" || var.client_secret == "") && var.identity_type == "UserAssigned" && try(length(var.identity_ids), 0) > 0)
       error_message = "When `kubelet_identity` is enabled - The `type` field in the `identity` block must be set to `UserAssigned` and `identity_ids` must be set."
     }
+    precondition {
+      condition     = var.brown_field_application_gateway_for_ingress == null || var.green_field_application_gateway_for_ingress == null
+      error_message = "Either one of `var.existing_application_gateway_for_ingress` or `var.new_application_gateway_for_ingress` must be `null`."
```
Comment (zioproto) [inline main.tf:576]: The error message needs to be updated with the new variable names.

File: variables.tf [hunk 6]
```diff
@@ -439,6 +459,21 @@ variable "enable_node_public_ip" {
   description = "(Optional) Should nodes in this Node Pool have a Public IP Address? Defaults to false."
 }
 
+variable "green_field_application_gateway_for_ingress" {
+  type = object({
+    name        = optional(string)
+    subnet_cidr = optional(string)
```
Comment (zioproto) [inline variables.tf:465]: For greenfield at least 1 of `subnet_id` or `subnet_cidr` should be provided. Should this be captured in a precondition ?

**Comment by lonegunmanb** [comment]
Hi @zioproto would you please give this pr another review? Thanks!

**Review by zioproto — APPROVED** [review]
